src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.Gss.cs

```diff
@@ -0,0 +1,336 @@
+            bool isEncrypt,
+            byte[] inputBytes,
+            int offset,
+            int count,
+            SafeGssBufferHandle outputMessageBuffer);
+
+        [DllImport(Interop.Libraries.SecurityNative)]
+        internal static extern Status GssUnwrap(
+            out Status minorStatus,
+            SafeGssContextHandle contextHandle,
+            byte[] inputBytes,
+            int offset,
+            int count,
+            SafeGssBufferHandle outputMessageBuffer);
+
+        [DllImport(Interop.Libraries.SecurityNative)]
+        internal static extern Status GssInquireSourceName(
+            out Status minorStatus,
+            SafeGssContextHandle contextHandle,
+            out SafeGssNameHandle srcName);
+
+        [DllImport(Interop.Libraries.SecurityNative)]
+        internal static extern Status GssDisplayName(
+            out Status minorStatus,
+            SafeGssNameHandle inputName,
+            SafeGssBufferHandle outputNameBuffer);
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct gss_buffer_desc
+        {
+            internal size_t length;
+            internal IntPtr value;
+        }
+
+        internal enum Status : uint
+        {
+            GSS_S_COMPLETE = 0,
+            GSS_S_CONTINUE_NEEDED = 1
+        }
+
+        [FlagsAttribute]
+        internal enum GssFlags : uint
+        {
+            GSS_C_DELEG_FLAG = 1,
+            GSS_C_MUTUAL_FLAG = 2,
+            GSS_C_REPLAY_FLAG = 4,
+            GSS_C_SEQUENCE_FLAG = 8,
+            GSS_C_CONF_FLAG = 16,
+            GSS_C_INTEG_FLAG = 32,
+            GSS_C_ANON_FLAG = 64,
+            GSS_C_PROT_READY_FLAG = 128,
+            GSS_C_TRANS_FLAG = 256,
+            GSS_C_DCE_STYLE = 4096,
+            GSS_C_IDENTIFY_FLAG = 8192,
+            GSS_C_EXTENDED_ERROR_FLAG = 16384,
+            GSS_C_DELEG_POLICY_FLAG = 32768
+        }
+    }
+}
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Wrapper around an output gss_buffer_desc*
+    /// </summary>
+    internal sealed class SafeGssBufferHandle : SafeHandle
+    {
+        private GCHandle _gch;
+
+        // Return the buffer size
+        public int Length
+        {
+            get
+            {
+                if (IsInvalid)
+                {
+                    return 0;
+                }
+                return (int)((Interop.libgssapi.gss_buffer_desc)_gch.Target).length;
+            }
+        }
+
+        // Return a pointer to where data resides
+        public IntPtr Value
+        {
+            get
+            {
+                if (IsInvalid)
+                {
+                    return IntPtr.Zero;
+                }
+                return ((Interop.libgssapi.gss_buffer_desc)_gch.Target).value;
+            }
+        }
+
+        public SafeGssBufferHandle() : base(IntPtr.Zero, true)
+        {
+            Interop.libgssapi.gss_buffer_desc buffer = new Interop.libgssapi.gss_buffer_desc
```

Comment:
Yes. With my suggestion [above] (https://github.com/dotnet/corefx/pull/5500#discussion_r49990601) we should be able to get rid of all this logic.